src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1731,8 +1731,8 @@ multi_join_restriction_hook(PlannerInfo *root,
 	joinRestriction->joinType = jointype;
 	joinRestriction->joinRestrictInfoList = restrictInfoList;
 	joinRestriction->plannerInfo = root;
-	joinRestriction->innerrel = innerrel;
-	joinRestriction->outerrel = outerrel;
+	joinRestriction->innerrelRelids = bms_copy(innerrel->relids);
```

Comment:
It looks suspicious. We still heavily rely on `plannerInfo` struct. Are we sure that it is not freed as well?

Reply:
Good point. I'll look into that. PlannerInfo doesn't look easily copyable as well.

Reply:
This code-path very critical in the sense that  we rely on it very heavily and it is called multiple times for every single query. 

Can you backtrace the code a bit to have a clear picture on what is happening when `geqo` is on/off? Why do we not get `innerrel/outerrel ` etc?

Also, I think no one experiencing #4223 means that we don't have any users relying on `geqo` at the moment. If the fix is not trivial, we can even consider ignoring `geqo` for the time being?

Reply:
`plannerInfo` I believe is a context struct that is allocated in the root memory context of the planner. geqo creates a new context that gets reset often. The call to this callback gets called with that context during geqo planning. inner and outer rel are set, but by the time we use the pointers the memory had been reset. Where it resets the memory it actually restores a bit of the state in the `plannerInfo` struct to not keep references to the reset memory.

Given this is the join hook I believe both innerrel and outerrel should be set. Would there be any possibility of it not being set? Did we have `NULL` checks where we were reading the relids before? In that case I'd like some `NULL` checks here instead before copying the relids. A `NULL` in relids is a valid bitmap with no bits set.

Reply:
>  A NULL in relids is a valid bitmap with no bits set.

That's a good point, `bms_is_empty()` returns true if the input bms is NULL.

Reply:
Flow of the code is as follows:
```
geqo
    geqo_eval
         CreateTempMemoryContext
                multi_join_restriction_hook
                      bms_copy
         DeleteTempMemoryContext
```

So, in a sense, we are copying a value that Postgres no longer considers valid as it deleted the corresponding memory context. 

Still,  we seem to  do the exact same thing for `extra->restrictlist`, and see https://github.com/citusdata/citus/pull/1349 for related discussions.  Given that the fields that we copy are pointing to `root`, we should be OK.


Should we maybe move the  deep copy parts next to each other?


```
diff --git a/src/backend/distributed/planner/distributed_planner.c b/src/backend/distributed/planner/distributed_planner.c
index a075a0f1e..035f2d157 100644
--- a/src/backend/distributed/planner/distributed_planner.c
+++ b/src/backend/distributed/planner/distributed_planner.c
@@ -1716,12 +1716,7 @@ multi_join_restriction_hook(PlannerInfo *root,
        MemoryContext restrictionsMemoryContext = plannerRestrictionContext->memoryContext;
        MemoryContext oldMemoryContext = MemoryContextSwitchTo(restrictionsMemoryContext);
 
-       /*
-        * We create a copy of restrictInfoList because it may be created in a memory
-        * context which will be deleted when we still need it, thus we create a copy
-        * of it in our memory context.
-        */
-       List *restrictInfoList = copyObject(extra->restrictlist);
+
 
        JoinRestrictionContext *joinRestrictionContext =
                plannerRestrictionContext->joinRestrictionContext;
@@ -1729,8 +1724,15 @@ multi_join_restriction_hook(PlannerInfo *root,
 
        JoinRestriction *joinRestriction = palloc0(sizeof(JoinRestriction));
        joinRestriction->joinType = jointype;
-       joinRestriction->joinRestrictInfoList = restrictInfoList;
        joinRestriction->plannerInfo = root;
+
+       /*
+        * We create a copy of restrictInfoList because it may be created in a memory
+        * context which will be deleted when we still need it, thus we create a copy
+        * of it in our memory context.
+        */
+       List *restrictInfoList = copyObject(extra->restrictlist);
+       joinRestriction->joinRestrictInfoList = restrictInfoList;
        joinRestriction->innerrelRelids = bms_copy(innerrel->relids);
        joinRestriction->outerrelRelids = bms_copy(outerrel->relids);

```

Reply:
I think moving the deep copies together and directly assigning them to the struct we keep makes the code more readable. I would suggest, if this doesn't break the line to go even further:

```C
joinRestriction->joinRestrictInfoList = copyObject(extra->restrictlist);
```

I have been fooled by the assignment of the `restrictInfoList` to the struct we keep before, while following the code.